csrc/scheduler/matmul_utils.cpp

```diff
@@ -0,0 +1,634 @@
+//!  layout) in problem in ProblemShape type.
+constexpr size_t N_POS = 1;
+//! A constant with position of K value (a number of rows in A tensor for TT
+//!  layout) in problem in ProblemShape type.
+constexpr size_t K_POS = 2;
+//! A constant with expected number of dimensions in ProblemShape type.
+constexpr size_t PROBLEM_DIMS = 3;
+
+// TODO: helpers to be moved to 'iter_visitor.h'
+std::deque<std::deque<Val*>> getAllDepndencyChains(
+    const std::vector<Val*>& producers,
+    const std::vector<Val*>& consumers) {
+  std::deque<std::deque<Val*>> all_paths;
+  for (auto* consumer : consumers) {
+    for (auto* producer : producers) {
+      auto paths = DependencyCheck::getAllDependencyChains(producer, consumer);
+      if (paths.empty()) {
+        continue;
+      }
+      all_paths.insert(
+          all_paths.end(),
+          std::make_move_iterator(paths.begin()),
+          std::make_move_iterator(paths.end()));
+    }
+  }
+
+  return all_paths;
+}
+
+//! A wrapper for printing debug details.
+void printMsg(const std::string& msg) {
+  std::cout << msg << std::endl;
+}
+
+//! A helper for deciding what kernel indexing mode use (int32_t or int64_t).
+//!  TODO: add strides to handle non-continous tensors
+PrimDataType getIndexType(const ProblemShape& problem_shape) {
+  // based on collectIndexMode function
+  constexpr int64_t most_positive_int32_index =
+      std::numeric_limits<int>::max() / 2;
+
+  const auto m = static_cast<int64_t>(problem_shape[M_POS]);
+  const auto n = static_cast<int64_t>(problem_shape[N_POS]);
+  const auto k = static_cast<int64_t>(problem_shape[K_POS]);
+
+  const bool use_i64_index = m * k > most_positive_int32_index || // tensor A
+      k * n > most_positive_int32_index || // tensor B
+      m * n > most_positive_int32_index; // output tensor
+
+  return use_i64_index ? PrimDataType::Int : PrimDataType::Int32;
+}
+
+//! A helper for deciding the type of MMA op for given fusion and problem shape.
+inline c10::optional<MmaOptions::MacroType> getMmaOp(
+    const int dev_version,
+    const ProblemShape& problem) {
+  using MacroType = MmaOptions::MacroType;
+
+  TORCH_INTERNAL_ASSERT(
+      problem.size() == PROBLEM_DIMS,
+      "Invalid size of problem shape (number of dimensions)");
+
+  // NOTE: A temp condition
+  const bool use_small_n =
+      ((problem[N_POS] % 8) == 0) && ((problem[N_POS] % 16) != 0);
+
+  switch (dev_version) {
+    case 70:
+      return MacroType::Volta_16_16_4;
+    case 75:
+      return (use_small_n) ? MacroType::Turing_16_8_16
+                           : MacroType::Turing_16_16_16;
+    case 80:
+      return (use_small_n) ? MacroType::Ampere_16_8_16
+                           : MacroType::Ampere_16_16_16;
+    default:
+      break;
+  }
+  return c10::nullopt;
+}
+
+//! A helper for checking if layout of MMA op's inputs. It will return optional
+//! message if check fails.
+LayoutData getInputsLayout(const MmaOp* mma_expr) {
```

Comment:
Note for future PR:
Does it make sense to make this a method of `MmaOp`?

Reply:
You are right, this can be moved as method of MmaOp. I will prepare PR with this change.

Thanks!